### PR TITLE
improves usability of style calls on axis=1 by automatically creating…

### PR DESCRIPTION
**Pull request by perryvais**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master | flake8 --diff`
- [ ] whatsnew entry

… the IndexSlice same as it does for axis=1

```
df = pd.DataFrame(index = ['x', 'y', 'z'])
df['a'] = [1,2,3]
df['b'] = [4,5,6]
df['c'] = [7,8,9]
df

df.style.background_gradient(subset='x', axis=1)
```
# otherwise this throws some vague exception. seems like if you pass in

axis this should just work

`df.style.background_gradient(subset=pd.IndexSlice['x',], axis=1)` #same
as this

### CODE REVIEW

**Comment by jreback** [comment]
- add a test [here](https://github.com/pydata/pandas/blob/master/pandas/tests/formats/test_style.py#L546)
- add a whatsnew note

**Comment by jreback** [comment]
cc @TomAugspurger

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -1925,7 +1925,7 @@ def maybe_droplevels(index, key):
     return index
 
 
-def _non_reducing_slice(slice_):
+def _non_reducing_slice(slice_, axis=0):
     """
     Ensurse that a slice doesn't reduce to a Series or Scalar.
```
Comment (jreback) [inline pandas/core/indexing.py]: add a doc-string up here. (and ``Ensurse`` is a typo)

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -1925,7 +1925,7 @@ def maybe_droplevels(index, key):
     return index
 
 
-def _non_reducing_slice(slice_):
+def _non_reducing_slice(slice_, axis=0):
```
Comment (jreback) [inline pandas/core/indexing.py]: prob also should pass ``self.data`` here, mainly for error checking / ndim (though this really only works for a DataFrame).

@TomAugspurger

**Comment by perryvais** [comment]
Fixed most of the issues you highlighted except for passing self.data. While that probably makes sense it is outside the scope of this bug and I didnt feel like i undertstood the code base enough to comfortably make it

File: doc/source/whatsnew/v0.18.2.txt [hunk 3]
```diff
@@ -259,3 +259,5 @@ Bug Fixes
 
 
 - Bug in ``Categorical.remove_unused_categories()`` changes ``.codes`` dtype to platform int (:issue:`13261`)
+
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.2.txt]: move to 0.19.0

File: pandas/tests/formats/test_style.py [hunk 4]
```diff
@@ -555,3 +555,12 @@ def test_background_gradient(self):
         result = (df.style.background_gradient(subset=pd.IndexSlice[1, 'A'])
                     ._compute().ctx)
         self.assertEqual(result[(1, 0)], ['background-color: #fff7fb'])
+
+        grad = df.style.background_gradient
```
Comment (jreback) [inline pandas/tests/formats/test_style.py]: add the issue number as a comment

**Comment by TomAugspurger** [comment]
@perryvais thanks for the PR, sorry it took so long for me to respond.

Is this closing an issue? I'd like to get some more background / use-cases on this.

One thing that worries me is this would spread the slicing/subsetting logic over two keyword arguments. On master, the logic is

- scalars and lists are treated as column slices
- tuples (e.g. output of IndexSlice) are passed to `.loc`

I viewed the case of `subset=column` special enough to warrant the special case instead of just saying "everything goes through `self.data.loc`", so pass in a valid indexer there.

With this PR you would also have to take into account the `axis` argument anytime you make a slice. Is this second special case common enough to justify the extra complexity? Was I wrong in my initial guess about `axis=0` being more common than `axis=1`?

**Comment by perryvais** [comment]
I don't think there is a preexisting issue but when i spoke to @jreback he believed it was a bug. I find the semantics of the styles to be a bit confusing but dont have a recommendation to clarify. The functionality is very helpful though. tried to create a gist but couldnt see styles but here is a screenshot. This PR gets the last line to behave like hte prior and consistent with the last line of the first cell.

image

**Comment by TomAugspurger** [comment]
Thanks, so to summarize.

## On Master:

- scalars and lists are treated as column slices
- tuples (e.g. output of IndexSlice) are passed to .loc

## This PR:

- scalars and lists are treated as
  + column slices with `axis=0`
  + index slices with `axis=1`
  + (what happens when there's ambiguity? i.e. `subset='x'` and there's a column and index label `'x'`)
- tuples (e.g. output of IndexSlice) are passed to .loc  (unchanged)

Anyone else have input here? I was thinking in a "tidy data" mindset when I gave scalar / list slices a special case instead of just always passing subset to `.loc`. That may have been a bad choice since these summary tables are often not tidy. Still I'm slightly -1 on this change since I don't like the slicing logic to be spread across two parameters, but willing to be convinced otherwise.

**Comment by perryvais** [comment]
Is ambiguity really an issue? default is axis=0 and that's pretty clear in sig. from PR

image

**Comment by TomAugspurger** [comment]
Right, as long as we define it to be that way there's no ambiguity. I removed that line.

**Comment by perryvais** [comment]
Sorry to nitpick but i think the slicing logic is already spread across 2 parameters. Although perhaps this is a separate bug or i'm doing something wrong.

IMHO this behavior is more intuitive and not ambiguous but i see your concerns. If anything we probably need to improve doc to show how to access via IndexSlice

image

**Comment by jreback** [comment]
ok with a setup like this (number slightly switched)
```
import pandas as pd
df = pd.DataFrame(index = ['x', 'y', 'z'])
df['a'] = [1,2,3]
df['b'] = [6,5,4]
df['c'] = [7,8,9]
```

image

The function is working correctly, but the API is too reminiscent of slicing and thus is ambiguous

```
def background_gradient(self, cmap='PuBu', low=0, high=0, axis=0,
                            subset=None):
        """
        Color the background in a gradient according to
        the data in each column (optionally row).
        Requires matplotlib.

        .. versionadded:: 0.17.1

        Parameters
        ----------
        cmap: str or colormap
            matplotlib colormap
        low, high: float
            compress the range by these values.
        axis: int or str
            1 or 'columns' for colunwise, 0 or 'index' for rowwise
        subset: IndexSlice
            a valid slice for ``data`` to limit the style application to
```
If ``axis`` instead becomes ``direction=left|top`` (e.g. ``left`` means ``left-to-right`` (and ``right`` -> ``right-to-left``), and ``top`` -> ``top-to-bottom``)

the its clear that ``subset`` controls slicing, and the ``direction`` controls gradient direction.

**Comment by jreback** [comment]
@TomAugspurger thoughts?

File: doc/source/whatsnew/v0.18.2.txt (hunk 3 above)
Comment (jreback) [inline doc/source/whatsnew/v0.18.2.txt]: move to 0.19.1

**Comment by jreback** [comment]
@perryvais pls rebase

**Comment by jreback** [comment]
@TomAugspurger comments?

**Comment by jreback** [comment]
closing as stale, but @TomAugspurger if you want to re-open / fixup. pls do so.

**Pull request closed without merging**